dds/DCPS/transport/rtps_udp/RtpsUdpReceiveStrategy.cpp

```diff
@@ -815,6 +814,38 @@ RtpsUdpReceiveStrategy::do_not_withhold_data_from(const RepoId& sub_id)
   readers_selected_.insert(sub_id);
 }
 
+// add readerID to readers_withheld_ and return true if a DirectedWrite message is not to this reader.
+bool RtpsUdpReceiveStrategy::withholdDirectedWrite(const RTPS::DataSubmessage& ds, const RepoId& readerID)
+{
+  bool directedWrite = false;
+  bool match = false;
+  for (CORBA::ULong i = 0; i < ds.inlineQos.length() && !match; ++i) {
+    if (ds.inlineQos[i]._d() == RTPS::PID_DIRECTED_WRITE) {
+      directedWrite = true;
+      if (ds.inlineQos[i].guid() == readerID) {
+        match = true;
+      }
+    }
+  }
+  if (directedWrite && !match) {
+    readers_withheld_.insert(readerID);
```

Comment:
`readers_withheld_`'s definition is not changing in this PR, so while we might want to change it I'd like to keep that for a different PR (unless we really need it here).

Reply:
The original code was like this without the 3-line if-statement:
    recvd_sample_ = &sample;
    readers_selected_.clear();
    readers_withheld_.clear();
    // If this sample should be withheld from some readers in order to maintain
    // in-order delivery, link_->received() will add it to readers_withheld_ otherwise
    // it will be added to readers_selected_
    link_->received(data, receiver_.source_guid_prefix_);
    if (readers_selected_.empty() && readers_withheld_.empty()) {
      listDirectedWriteReaders(data);
    }
    recvd_sample_ = 0;

I had the same feeling that not only the two lists readers_withheld_ and readers_selected_, but also recvd_sample_ were used more like local variables instead of instance variables. Notice that recvd_sample_ is set to &sample right before clearing the two lists and then set to 0 right after the call to link_->received(). I think this is because of the closely tied relationship between strategy and datalink objects. That is why link_->received() can fill the two lists without even taking them as mutable references (it could not do that if they were local variables).

I also agree with Adam that it's best to change this tangled relationship in a separate PR, because that will affect all the other delivery strategies, not just the DirectedWrite. 

The real reason that best effort readers did not get the appropriate strategies applied to them is that the current code does not fill the readers_of_writer_ list in RtpsUdpDataLink::associated(). Strategies like duplicate seq# are still not applied to best effort readers. There should be a better way than to tie association with receiving strategies, because this does not extend well. Adding a new strategy that needs to work for both, such DirectedWrite, will have problems.  

I had another version of directedWrite code:
    if (processDirectedWrite(sample, data)) {
      break;
    }
The processDirectedWrite() method did not use the two members (readers_withheld_ and readers_selected_). It retrieved the directedWrite guids into a local variable, and call link_->data_received_include(sample, readers);
This version worked for best effort readers, but hung on reliable readers.

Now we have a way to process DirectedWrite messages for both reliable and best effort readers, and a test that can successfully verify for both situations. I'd like to use a separate PR to separate the responsibilities between the strategy and the datalink objects.

Reply:
In this PR, please see what needs to happen to remove the call from `RtpsUdpDataLink::RtpsReader` to `RtpsUdpReceiveStrategy`.

The data link's responsibility is to implement the reliability protocol for rtps_udp.  It shouldn't need to be involved in directed write.